src/shims/foreign_items.rs

```diff
@@ -627,8 +606,8 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
             // Platform-specific shims
             _ => match this.tcx.sess.target.os.as_str() {
-                "linux" | "macos" => return shims::posix::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, abi, args, dest, ret),
-                "windows" => return shims::windows::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, abi, args, dest, ret),
+                "linux" | "macos" => return shims::posix::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, link_name_sym, abi, args, dest, ret),
+                "windows" => return shims::windows::foreign_items::EvalContextExt::emulate_foreign_item_by_name(this, link_name, link_name_sym, abi, args, dest, ret),
```

Comment:
Seems a bit annoying that we pass both `link_name` and `link_name_sym` here -- can we avoid passing the string?
(Maybe we can even avoid the `link_name` variable entirely, so we can drop the `_sym` suffix?)

Reply:
That means we need to call `Symbol::as_str` more than once for the same `Symbol` (in different functions) or intern link names of every shim we supported, which sounds inefficient...

Reply:
I was thinking we would do the former. Is `Symbol::as_str()` that slow? Given that this only happens on calls to foreign items, I don't think a slight perf hit will even be measurable.

Reply:
I tried to make the change, but now I don't know how to deal with this:
https://github.com/rust-lang/miri/blob/d39f0c64b8b369188a73a655716ab56683a6537b/src/shims/foreign_items.rs#L222-L223
The `link_name` passed to `check_abi_and_shim_symbol_clash` needs to be the *actual* link name -- that is, the name with suffix.

Reply:
> EDIT: I think I can copy that `trim_end_matches` to (or introduce a helper function that calls `trim_end_matches` and call it at) every `match`.

Done, although "32-bit macOS" (`i686-apple-darwin`) seems broken on `master` at this moment:
<details>

```
diff of stdout:

-hello dup fd
-

The actual stdout differed from the expected stdout.
Actual stdout saved to /tmp/compiletestklptUY/fs.stage-id.stdout
diff of stderr:

-hello dup fd
+thread 'main' panicked at 'assertion failed: `(left == right)`
+  left: `60129542144`,
+ right: `14`', $DIR/fs.rs:82:5
+stack backtrace:
+   0: core::panicking::panic_fmt::panic_impl
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panicking.rs:515:5
+   1: core::panicking::panic_fmt
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/panicking.rs:92:14
+   2: core::panicking::assert_failed_inner
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/panicking.rs:160:17
+   3: core::panicking::assert_failed::<u64, u64>
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/panicking.rs:117:5
+   4: test_file
+             at $DIR/fs.rs:82:5
+   5: main
+             at $DIR/fs.rs:17:5
+   6: <fn() as std::ops::FnOnce<()>>::call_once - shim(fn())
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
+   7: std::rt::lang_start::<()>::{closure#0}
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/rt.rs:49:18
+   8: std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/ops/function.rs:259:13
+   9: std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panicking.rs:401:40
+  10: std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panicking.rs:365:19
+  11: std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panic.rs:434:14
+  12: std::rt::lang_start_internal
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/rt.rs:34:21
+  13: std::rt::lang_start::<()>
+             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/rt.rs:48:5
+note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
 

The actual stderr differed from the expected stderr.
Actual stderr saved to /tmp/compiletestklptUY/fs.stage-id.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args fs.rs`

error: 2 errors occurred comparing output.
status: exit status: 101
command: "/home/hyd-dev/Code/rust-lang/miri/target/x86_64-unknown-linux-gnu/release/miri" "tests/run-pass/fs.rs" "-L" "/tmp/compiletestklptUY" "--target=i686-apple-darwin" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestklptUY/fs.stage-id" "-A" "unused" "--edition" "2018" "-Dwarnings" "-Dunused" "--sysroot" "/home/hyd-dev/.cache/miri" "-Zmiri-disable-isolation" "-L" "/tmp/compiletestklptUY/fs.stage-id.aux"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `60129542144`,
 right: `14`', tests/run-pass/fs.rs:82:5
stack backtrace:
   0: core::panicking::panic_fmt::panic_impl
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/panicking.rs:92:14
   2: core::panicking::assert_failed_inner
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/panicking.rs:160:17
   3: core::panicking::assert_failed::<u64, u64>
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/panicking.rs:117:5
   4: test_file
             at tests/run-pass/fs.rs:82:5
   5: main
             at tests/run-pass/fs.rs:17:5
   6: <fn() as std::ops::FnOnce<()>>::call_once - shim(fn())
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
   7: std::rt::lang_start::<()>::{closure#0}
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/rt.rs:49:18
   8: std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/ops/function.rs:259:13
   9: std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panicking.rs:401:40
  10: std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panicking.rs:365:19
  11: std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/panic.rs:434:14
  12: std::rt::lang_start_internal
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/rt.rs:34:21
  13: std::rt::lang_start::<()>
             at /home/hyd-dev/.rustup/toolchains/miri/lib/rustlib/src/rust/library/std/src/rt.rs:48:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

------------------------------------------

test [ui] run-pass/fs.rs ... FAILED
```
</details>

Reply:
> EDIT: I think I can copy that trim_end_matches to (or introduce a helper function that calls trim_end_matches and call it at) every match.

Only some of the `match` should actually need it. I forgot how many of the functions actually need this...

> "32-bit macOS"

FAIK Apple doesn't support this target any more, so neither do we.

Reply:
32bit macOS targets [are tier 3](https://blog.rust-lang.org/2020/01/03/reducing-support-for-32-bit-apple-targets.html).

So if this is only needed for them... then as far as I am concerned we can just entirely remove this suffix handling.